plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Auction.hs

```diff
@@ -0,0 +1,269 @@
+currentState client = SM.getOnChainState client >>= \case
+    Just ((TypedScriptTxOut{tyTxOutData=Ongoing s}, _), _) -> do
+        logInfo (CurrentState s)
+        pure (Just s)
+    _ -> do
+        logWarn CurrentStateNotFound
+        pure Nothing
+
+{- Note [Buyer client]
+
+In the buyer client we want to keep track of the on-chain state of the auction
+to give our user a chance to react if they are outbid by somebody else.
+
+At the same time we want to have the "bid" endpoint active for any bids of our
+own, and we want to stop the client when the auction is over.
+
+To achieve this, we have a loop where we wait for one of several events to
+happen and then deal with the event. The waiting is implemented in
+@waitForChange@ and the event handling is in @handleChange@.
+
+Updates to the user are provided via the log functionality. This is not as bad
+as it sounds because the log records JSON objects so we are at least semi-
+structured. (Basically we have a @Writer [Aeson.Value]@ - in the future we
+should just make it a @Writer a@ for some user-defined @a@ so that we can
+export the "current state" of the app more easily.)
+
+-}
+
+data BuyerChange =
+        AuctionIsOver HighestBid -- ^ The auction has ended with the highest bid
+        | SubmitOwnBid Ada -- ^ We want to submit a new bid
+        | OtherBid HighestBid -- ^ Another buyer submitted a higher bid
+        | NoChange HighestBid -- ^ Nothing has changed
+
+waitForChange :: AuctionParams -> StateMachineClient AuctionState AuctionInput -> HighestBid -> Contract BuyerSchema SM.SMContractError BuyerChange
+waitForChange AuctionParams{apEndTime} client lastHighestBid = do
+    s <- currentSlot
+    let
+        auctionOver = awaitSlot apEndTime >> pure (AuctionIsOver lastHighestBid)
+        submitOwnBid = SubmitOwnBid <$> endpoint @"bid"
+        otherBid = do
+            let address = Scripts.scriptAddress (validatorInstance (SM.scInstance client))
+                targetSlot = succ (succ s) -- FIXME (jm): There is some off-by-one thing going on that requires us to
+                                           -- use succ.succ instead of just a single succ if we want 'addressChangeRequest'
+                                           -- to wait for the next slot to begin.
+                                           -- I don't have the time to look into that atm though :(
+            AddressChangeResponse{acrTxns} <- addressChangeRequest AddressChangeRequest{acreqSlot=targetSlot, acreqAddress = address}
+            case acrTxns of
+                [] -> pure (NoChange lastHighestBid)
+                _  -> currentState client >>= pure . maybe (AuctionIsOver lastHighestBid) OtherBid
+
+    -- see note [Buyer client]
+    auctionOver `select` submitOwnBid `select` otherBid
+
+handleChange :: StateMachineClient AuctionState AuctionInput -> HighestBid -> BuyerChange -> Contract BuyerSchema SM.SMContractError (Either HighestBid ())
```

Comment:
`handleEvent`?

Reply:
I guess it's not totally clear to me why we don't just fuse this function with the previous one, so that we just do the thing that we're supposed to do after the condition holds? i.e. just stick `>> stop` at the end of line 226, etc

Reply:
I thought it would be easier to understand that way, if the waiting and handling are separated.